TINKERPOP-2813 Make gremlin-driver more optimistic about host availability

spmallette opened the pull request:
https://issues.apache.org/jira/browse/TINKERPOP-2813

Summarized the major points of the change in the Upgrade Docs. Might be best to start there if reviewing this. Made some other adjustments to log better data for debugging purposes with the `ConnectionPool`.  Ended up making a number of adjustments to the thread pools to which various jobs were being submitted. Untangling that led to the addition of two pools which helped bring back parallel host initialization and got rid of random usage of the fork/join pool. I sense that something could be done to eliminate some of these pools but I believe it will come at the cost of a much larger refactoring effort to untangle how some of these jobs get scheduled up. I dunno.....not super pleased with it but gotta go pencils down at some point.

there've been many days of continuous testing on these changes at this point from various folks and all results have been favorable. calling out @kenhuuu specifically, he did some basic performance testing and profiling and didn't find any problems to note:

```text
Test Case:
    2450 short queries (return in milliseconds)
    1280 mix of short and medium queries

Setup:
    m5.2xlarge
    3.5.4 Gremlin Server with gremlin-server.yaml
    Defaults for connection pooling and in flight per connection settings for Java driver

Results for Three Run Average.
    3.5.4:
        Short: 29912ms
        Mixed: 65689ms

    TINKERPOP-2813 branch:
        Short: 30161ms
        Mixed: 65422ms
```

vkagamlyk commented on gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Client.java:
```diff
@@ -495,49 +496,45 @@ public Client alias(final Map<String, String> aliases) {
-            final Host bestHost = possibleHosts.next();
+            // try a random host if none are marked available. maybe it will reconnect in the meantime. better than
+            // going straight to a fast NoHostAvailableException as was the case in versions 3.5.4 and earlier
+            final Host bestHost = possibleHosts.hasNext() ? possibleHosts.next() : chooseRandomHost();
```
For example we have a cluster of 3 instances.
For some reason (reconfiguration, scaling, network fluke) all instances restarted.
After some time can we connect to one random host and use it, but all the others will remain disconnected.
Maybe it makes sense to check all unavailable hosts when the user sends a request, but not more than once every 10-60 seconds?

FlorianHockmann commented on docs/src/reference/gremlin-variants.asciidoc:
```diff
@@ -658,6 +658,31 @@ social.persons("marko").knows("josh");
+1. The server has actually reached its maximum capacity or the driver has just learned that the server is unreachable.
+2. The client is throttling requests.
+
+The latter of the two can be addressed from the driver side in the following ways:
```
Not sure if that's necessary at all, but how about mentioning that the exception message / the error log message should contain information about the state of the `ConnectionPool` and its `Connections` to understand which of the two cases caused the problem?

FlorianHockmann commented on docs/src/reference/gremlin-variants.asciidoc:
```diff
@@ -658,6 +658,31 @@ social.persons("marko").knows("josh");
+where this occurs:
+
+1. The server has actually reached its maximum capacity or the driver has just learned that the server is unreachable.
+2. The client is throttling requests.
```
To be honest, when I just read this once it was not directly clear to me what that means. Why is it throttling requests?
Simply adding "[...] because the pool is exhausted" would already help.

FlorianHockmann commented on gremlin-server/src/test/java/org/apache/tinkerpop/gremlin/driver/ClientConnectionIntegrateTest.java:
```diff
@@ -110,4 +124,100 @@ public void shouldCloseConnectionDeadDueToUnRecoverableError() throws Exception
+        }, "worker-shouldSucceedWithJitteryConnection").start();
+
+        // wait long enough for the jitters to kick in at least a little
+        while (latch.getCount() > 500) {
```
(nitpick) Is this `500` because then half the requests have been executed? In that case, I'd change this to `requests / 2` to make this clear.

FlorianHockmann commented on docs/src/upgrade/release-3.5.x.asciidoc:
```diff
@@ -50,6 +50,48 @@ var client = new GremlinClient(new GremlinServer("localhost", 8182), loggerFacto
+in an orderly fashion so as to allow for existing requests to complete if at all possible, it immediately blocks new
+requests producing an immediate `NoHostAvailableException` (assuming there are actually no other hosts to route
+requests to). There is then some delay to reconnect to the server and re-establish each of the connections in the pool,
+which might take some time especially if there was a large pool. If an application were sending hundreds of request
```
```suggestion
which might take some time especially if there was a large pool. If an application were sending hundreds of requests
```

FlorianHockmann reviewed (commented):
Just reviewed this, mostly to better understand the improvement you're making to the driver here myself.

Added some inline comments, but they are just nits or questions I had.

kenhuuu commented:
LGTM. I've tested this and it greatly improves the handling of temporary errors.

vkagamlyk commented on gremlin-server/src/test/java/org/apache/tinkerpop/gremlin/driver/ClientConnectionIntegrateTest.java:
```diff
@@ -21,20 +21,34 @@
 import static org.junit.Assert.fail;
 
 public class ClientConnectionIntegrateTest extends AbstractGremlinServerIntegrationTest {
+    private static final org.slf4j.Logger logger = LoggerFactory.getLogger(ClientConnectionIntegrateTest.class);
```
looks redundant, tests use log4j

vkagamlyk commented on gremlin-server/src/test/java/org/apache/tinkerpop/gremlin/driver/ClientConnectionIntegrateTest.java:
```diff
@@ -110,4 +124,100 @@ public void shouldCloseConnectionDeadDueToUnRecoverableError() throws Exception
+
+        // if there was a exception in the worker thread, then it had better be a TimeoutException
+        assertThat(hadFailOtherThanTimeout.get(), is(false));
+
+        connectionFactory.jittery = false;
```
This `connectionFactory` is not used more in this scope
```suggestion
```

spmallette commented:
@FlorianHockmann Thanks for the review. I addressed your points. More refinements to the documentation are coming I think, as @kenhuuu will be adding a follow-on PR with more tests. I'll follow up that with a few more documentation adjustments.
